File: code/code39.py
from typing import List
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        arr = []
        def dfs(start: int,tgt:int):
            if tgt == 0:
                ans.append(arr.copy())
                return
            if tgt < 0:
                return
            for index,candidate in enumerate(candidates):
                if index < start:
                    continue
                arr.append(candidate)
                dfs(index,tgt - candidate)
                arr.pop()
            return
        dfs(0,target)
        return ans
if __name__ == "__main__":
    candidates = [2,3,6,7]
    solution = Solution()
    ans = solution.combinationSum(candidates,7)
    print(ans)

File: code/code39.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    vector<vector<int>> ans;
public:
    void dfs(vector<int>& candidates,vector<int> temp,int sum,int target,int start) {
        if (sum > target) return;
        if (sum == target) {
            ans.push_back(temp);
        }
        for (auto i = start;i < candidates.size();i++) {
            temp.push_back(candidates[i]);
            dfs(candidates,temp,sum+candidates[i],target,i);
            temp.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        dfs(candidates,temp,0,target,0);
        return ans;
    }
};